### PR TITLE
Improve Mac package for enterprise install scenarios

### PR DESCRIPTION
**Pull request by nstrauss**
Thank you for investing time into a package! This PR suggests a few improvements to use with large scale enterprise deployments.

The Mac package as is assumes there's a logged in console user in order to set permissions on `/opt/homebrew` and copy cache files to `~/Library/Caches/Homebrew/api`. Usually when a person installs the package manually. However, in an enterprise environment that's often not the case.

For organizationally managed Macs using software management tools like Jamf or [Munki](https://github.com/munki/munk) which initiate installs, it's common for there to be no logged in user or that a user hasn't even been created yet during install. These solutions and similar often run at the login window via a launch daemon before a valid console user has logged in. It's also possible to use Apple's Automated Device Enrollment to deliver packages via MDM before setup assistant is complete. Introducing fallbacks in cases where there's no active console user should cover most enterprise focused edge cases.

- When a known system account such `loginwindow` is the current console user, infer the intended user by looking at the most recently created UID over 500.
- Support passing in a `HOMEBREW_PKG_USER` environment variable for admins to have better control over the target user where required (a lab environment). For example - `HOMEBREW_PKG_USER="nathaniel.strauss" installer -pkg Homebrew-4.1.15 -target /`
- Fix moving `/opt/homebrew/cache_api` to user directory. This previously did not work.
- Bump package required minimum OS to 12.0 to match Homebrew.
- Remove sudo from postinstall script. The installer already runs as root.
- Remove installing `gh` from workflow. GitHub CLI is baked into the runner. https://github.com/actions/runner-images/blob/375e9a1c270ec81a3ffd5b429dcdee9508b46124/images/macos/macos-13-Readme.md?plain=1#L57

### CODE REVIEW

File: package/scripts/postinstall [hunk 1]
```diff
@@ -43,20 +43,42 @@ then
...
-mv -v "${homebrew_directory}/cache_api/"* "${user_api_cache_dir}"
-rm -vrf "${homebrew_directory}/cache_api"
+if [[ ! "${logged_in_user}" == "root" ]]
+then
+  user_home_dir=$(dscl . read /Users/"${logged_in_user}" NFSHomeDirectory | awk '{ print $2 }')
+  user_cache_dir="${user_home_dir}/Library/Caches/Homebrew"
+  user_api_cache_dir="${user_cache_dir}/api"
+  mkdir -vp "${user_api_cache_dir}"
+  mv -v "${homebrew_directory}/cache_api/"* "${user_api_cache_dir}"
+  chown -R "${logged_in_user}:staff" "${user_cache_dir}"
```
Comment (MikeMcQuaid) [inline package/scripts/postinstall:82]: You're using `staff` in some places and `admin` in others.

File: package/Distribution.xml [hunk 2]
```diff
@@ -4,7 +4,7 @@
   <options customize="never" hostArchitectures="x86_64,arm64" rootVolumeOnly="true"/>
   <volume-check>
     <allowed-os-versions>
-        <os-version min="11.0.0"/>
+        <os-version min="12.0.0"/>
```
Comment (MikeMcQuaid) [inline package/Distribution.xml:7]: Good catch! Not your mistake to fix but, if you're feeling generous, could you add a link to this file somewhere around here: https://github.com/Homebrew/brew/blob/d0eef12ea40ebb8e72b708336f8aa6d21d955061/Library/Homebrew/brew.sh#L507-L513

so this gets updated for future versions.

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks @nstrauss, great work so far ❤️! A bunch of comments here but I think we're pretty close. The TL;DR is that I think we want to avoid as many edge-cases here as possible. I think `HOMEBREW_PKG_USER` is a good addition but trying to guess the correct user or ending up with a `root`-owned installation gets a fairly big 👎🏻 from me, I'm afraid.

**Comment by MikeMcQuaid** [comment]
> * Remove installing `gh` from workflow. GitHub CLI is baked into the runner. https://github.com/actions/runner-images/blob/375e9a1c270ec81a3ffd5b429dcdee9508b46124/images/macos/macos-13-Readme.md?plain=1#L57

We definitely have seen cases where it's not been present 🤷🏻. Would rather just leave this as it guarantees we have the newest version and adds minimal overhead.

> For organizationally managed Macs using software management tools like Jamf or [Munki](https://github.com/munki/munk) which initiate installs, it's common for there to be no logged in user or that a user hasn't even been created yet during install. These solutions and similar often run at the login window via a launch daemon before a valid console user has logged in. It's also possible to use Apple's Automated Device Enrollment to deliver packages via MDM before setup assistant is complete. Introducing fallbacks in cases where there's no active console user should cover most enterprise focused edge cases.

I'm game for some additional configuration e.g. `HOMEBREW_PKG_USER` but a fall-back that results in a `root`-owned Homebrew or guesses at the correct user doesn't seem like an acceptable fall-back, I'm afraid.

File: package/scripts/postinstall [hunk 3]
```diff
@@ -32,7 +32,7 @@ git config --global --unset safe.directory "${homebrew_directory}"
 # move to /usr/local if on x86_64
 if [[ $(uname -m) == "x86_64" ]]
 then
-  sudo mkdir -vp /usr/local/bin /usr/local/Homebrew
+  mkdir -vp /usr/local/bin /usr/local/Homebrew
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:35]: I'd prefer the script be invoked with `sudo` while testing, but it's a matter of opinion/style and I don't feel too strongly about it. Adding `sudo` back and including in the `chown -R ~/Library/Caches` line as well for the same reason.

File: package/scripts/postinstall (hunk 1 above)
Comment (nstrauss) [reply package/scripts/postinstall:82]: As is the script gets a user's group, which is almost always `staff`, and then sets ownership for `/opt/homebrew` and `~/Library/Caches/Homebrew`. I won't call the behavior incorrect, but it doesn't mimic what the Homebrew install script does. My goal was to match permissions for a typical install. I did check my test machines to ensure this was accurate across Intel and Apple silicon. 

`/opt/homebrew` should be `$username:admin`, `~/Library/Caches/Homebrew` (like with everything in `~/Library/Caches` should be `$username:staff`.

File: package/scripts/postinstall [hunk 4]
```diff
@@ -43,20 +43,42 @@ then
...
 
-# set permissions
+# get valid logged in user
 logged_in_user=$(echo "show State:/Users/ConsoleUser" | scutil | awk '/Name :/ { print $3 }')
-group=$(id -gn "${logged_in_user}")
+
+# optionally pass an environment variable to `installer` to define user
+if [[ -n "${HOMEBREW_PKG_USER:-}" ]]
+then
+  logged_in_user="${HOMEBREW_PKG_USER}"
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:54]: Good point. Added to Installation.md.

File: package/scripts/postinstall [hunk 5]
```diff
@@ -43,20 +43,42 @@ then
...
 fi
 
 # move API cache to ~/Library/Caches/Homebrew
-user_api_cache_dir=~"${logged_in_user}"/Library/Caches/Homebrew/api
-mkdir -vp "${user_api_cache_dir}"
-mv -v "${homebrew_directory}/cache_api/"* "${user_api_cache_dir}"
-rm -vrf "${homebrew_directory}/cache_api"
+if [[ ! "${logged_in_user}" == "root" ]]
+then
+  user_home_dir=$(dscl . read /Users/"${logged_in_user}" NFSHomeDirectory | awk '{ print $2 }')
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:77]: It does not. It makes a literal `~nathaniel.strauss` directory in `/opt/homebrew`.

File: package/scripts/postinstall [hunk 6]
```diff
@@ -43,20 +43,42 @@ then
...
-group=$(id -gn "${logged_in_user}")
+
+# optionally pass an environment variable to `installer` to define user
+if [[ -n "${HOMEBREW_PKG_USER:-}" ]]
+then
+  logged_in_user="${HOMEBREW_PKG_USER}"
+elif [[ "${logged_in_user}" =~ _mbsetupuser|loginwindow|root ]] || [[ -z "${logged_in_user}" ]]
+then
+  # get last created account UID over 500
+  logged_in_user=$(dscl . list /Users UniqueID | awk '$2 > 500 { print $1 }' | tail -n 1)
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:58]: Though I agree it's brittle to a degree, this is the crux of the PR and without it or a similar approach makes my idea is somewhat useless. My main goal is to support successful install without a logged in user. Though there will definitely be differences of opinion, for this to be possible, we have to compromise on a strategy. The options as I see them are...

1. Get the last created UID over 500.
2. Get the first created UID over 500.
3. Get the last logged in user.
4. Specify a user (`HOMEBREW_PKG_USER`).

Not everyone will agree. Right now however there's no support at all. When Homebrew.pkg is installed at the login window with no console user, it's owned by `root` or another system account. Considering the vast majority of enterprise managed Macs have a 1:1 relationship with an assigned user, it's generally safe to say UID 501 will be the intended owner. Using `dscl` here is a way to make that a bit more flexible.

File: package/scripts/postinstall [hunk 7]
```diff
@@ -43,20 +43,42 @@ then
...
+if [[ ! "${logged_in_user}" == "root" ]]
+then
+  user_home_dir=$(dscl . read /Users/"${logged_in_user}" NFSHomeDirectory | awk '{ print $2 }')
+  user_cache_dir="${user_home_dir}/Library/Caches/Homebrew"
+  user_api_cache_dir="${user_cache_dir}/api"
+  mkdir -vp "${user_api_cache_dir}"
+  mv -v "${homebrew_directory}/cache_api/"* "${user_api_cache_dir}"
+  chown -R "${logged_in_user}:staff" "${user_cache_dir}"
+  rm -vrf "${homebrew_directory}/cache_api"
+fi
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:84]: Yeah, agreed. The idea of `root` owning Homebrew was my mistake. I was trying to find a way to not make the install fail entirely.

File: package/scripts/postinstall [hunk 8]
```diff
@@ -43,20 +43,42 @@ then
...
+  logged_in_user="${HOMEBREW_PKG_USER}"
+elif [[ "${logged_in_user}" =~ _mbsetupuser|loginwindow|root ]] || [[ -z "${logged_in_user}" ]]
+then
+  # get last created account UID over 500
+  logged_in_user=$(dscl . list /Users UniqueID | awk '$2 > 500 { print $1 }' | tail -n 1)
+  # no valid accounts exist yet, use root temporarily
+  if [[ -z "${logged_in_user}" ]]
+  then
+    logged_in_user="root"
+  fi
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:63]: Agreed assigning `root` as owner is not the solution here. How would we like the install to fail? At this point in the install files are already on disk in `/opt/homebrew`. If using this conditional to bail halfway through users might assume the package "failed successfully" or is in a healthy enough state to proceed. Or Homebrew could be broken enough they don't know what to do. 

I think the best solution is to fail early. Since we're in agreement installs with no user created yet shouldn't be supported, let's move or add `logged_in_user` logic to a preinstall script. Then a precondition check can fail before the package lays down files. Thoughts?

```bash
  # no valid accounts exist yet, use root temporarily
  if [[ -z "${logged_in_user}" ]]
  then
    logged_in_user="root"
  fi
```

The above would change to something like...

```bash
  # no valid accounts exist yet, use root temporarily
  if [[ -z "${logged_in_user}" ]]
  then
    echo "No valid user to install Homebrew. Exiting."
    exit 1
  fi
```

https://github.com/Homebrew/brew/blob/709b4680e76ea9fddcf4eee0b869dc959de07245/package/scripts/postinstall#L51-L64

File: package/scripts/postinstall [hunk 9]
```diff
@@ -43,20 +43,42 @@ then
 fi
 
 # create missing directories
-sudo mkdir -vp Cellar Frameworks etc include lib opt sbin share var/homebrew/linked
+mkdir -vp Cellar Frameworks etc include lib opt sbin share var/homebrew/linked
 
-# set permissions
+# get valid logged in user
 logged_in_user=$(echo "show State:/Users/ConsoleUser" | scutil | awk '/Name :/ { print $3 }')
```
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall:49]: Matter of style and means one less `elif`. Getting the current console user is the main method, instantiate as such. Then if other conditions are met reassign.

File: package/Distribution.xml (hunk 2 above)
Comment (nstrauss) [reply package/Distribution.xml:7]: I see you've done this already in https://github.com/Homebrew/brew/pull/16077 so going to keep the work there.

**Comment by nstrauss** [comment]
@MikeMcQuaid Thank you for comments and feedback 🙏 Made a few changes in response, but would like more direction on a few other items.

File: package/scripts/postinstall (hunk 6 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply package/scripts/postinstall:58]: > 4\. Specify a user (`HOMEBREW_PKG_USER`).

This seems the best option here.

> Not everyone will agree.

Yup, but that doesn't mean Homebrew needs to offer multiple options 😁 

> When Homebrew.pkg is installed at the login window with no console user, it's owned by `root` or another system account.

Yeh, this should be considered a bug. I agree with detecting these cases and handling it, I just think that we should fail if either a user isn't specified or logged-in.

> Considering the vast majority of enterprise managed Macs have a 1:1 relationship with an assigned user, it's generally safe to say UID 501 will be the intended owner.

This feels too much like unexpected magic to me. Requiring an explicit user feels like a better fit here.

File: package/scripts/postinstall (hunk 8 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply package/scripts/postinstall:63]: > Since we're in agreement installs with no user created yet shouldn't be supported, let's move or add `logged_in_user` logic to a preinstall script.

Agreed 👍🏻 

I think a `Distribution.xml` `<script>` check, if possible, may be even better as it allows for a nicer error message flow.

File: package/scripts/postinstall (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply package/scripts/postinstall:49]: It's nicer (and less error-prone) to avoid shelling out in cases we know we won't need the result.

File: package/scripts/postinstall (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply package/scripts/postinstall:54]: Looks good, thanks!

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the changes here, good progress!

**Comment by MikeMcQuaid** [comment]
@nstrauss I merged https://github.com/Homebrew/brew/pull/16077 so I've merged that in here to resolve merge conflicts for you. A few related notes:
- you changed my mind on `sudo` so I've removed all of those (plus the one I got you to 😅)
- I removed the unnecessary shell out if `HOMEBREW_PKG_USER` is passed

Feels like the next step, as discussed would be to:
- check for `HOMEBREW_PKG_USER` or a logged-in user in `Distribution.xml` or `preinstall` (depending on what can get the nicest user experience on failure)

**Comment by MikeMcQuaid** [comment]
> * Support passing in a `HOMEBREW_PKG_USER` environment variable for admins to have better control over the target user where required (a lab environment). For example - `HOMEBREW_PKG_USER="nathaniel.strauss" installer -pkg Homebrew-4.1.15 -target /`

@nstrauss I was testing this out and couldn't seem to get it to work. https://stackoverflow.com/questions/55511165/pass-variables-to-the-installer-on-osx-package-installer makes it look like it may be more involved, unfortunately.

**Comment by nstrauss** [comment]
> > * Support passing in a `HOMEBREW_PKG_USER` environment variable for admins to have better control over the target user where required (a lab environment). For example - `HOMEBREW_PKG_USER="nathaniel.strauss" installer -pkg Homebrew-4.1.15 -target /`
> 
> @nstrauss I was testing this out and couldn't seem to get it to work. https://stackoverflow.com/questions/55511165/pass-variables-to-the-installer-on-osx-package-installer makes it look like it may be more involved, unfortunately.

Tested this myself and the linked article is correct. No longer working as I thought. May have another idea.

File: package/scripts/postinstall (hunk 8 above)
[the remark this answers is not in the document]
Comment (nstrauss) [reply package/scripts/postinstall]: Tried to shell out in `Distribution.xml` using `run` (https://developer.apple.com/documentation/installer_js/system/1812364-run), but Apple has deprecated it in `productbuild`. Instead I opted for a preinstall script. Though the user experience isn't quite as nice, this error should never be encountered by a logged in user installing the package manually. The other option is to use `users.desktopSessionCount` (https://developer.apple.com/documentation/installer_js/system/1812277-users_desktopsessioncount) to determine if a user is logged in, but still holding out hope we can reintroduce login window support.

```
productbuild: [WARNING]: Specifying [allow-external-scripts='true'] is deprecated.
           Having this in your distribution will break the installability
           of this product in an upcoming release of macOS.
           Apple reserves the right to reject notarization of any product with this option.
```
Screenshot 2023-10-04 at 12 37 46 PM

**Comment by nstrauss** [comment]
@MikeMcQuaid Alright, thanks for bearing with me. Here's where we're at.

- Permissions improvements to postinstall.
- A new preinstall to hard fail early when there is no logged in user.
- The `HOMEBREW_PKG_USER` environment variable doesn't work, and probably won't without hacky workarounds. Going to keep thinking about alternatives.

So slightly dismayed the PR has been hollowed out from its original intent. My main goal is still to get login window installations working. Your preference seems to be to explicitly define a user when one is not logged in. I respect that approach as it means admins can bring their own logic for which user to specify instead of Homebrew being heavily opinionated by shelling out to `dscl`. I'm trying to think up ways to do so outside the non-functional env variable. Open to ideas there. Is there a canonical place Homebrew stores settings?

I'm still of the opinion falling back to first/last user created UID isn't all that bad, but it does mean Homebrew is taking a position, and with no configurable way to define user could end up with unexpected behavior.

**Comment by nstrauss** [comment]
I think I have a better solution in place. https://github.com/Homebrew/brew/pull/16073/commits/2c5788178f8ce56578253e2b9dcaef933b8efe69 adds support for reading a defined user from `/var/tmp/.homebrew_pkg_user.plist`. I'm open to changing the location and format as long as it exists outside any Homebrew directory. I chose a property list file since it's a commonly used structured data format in the Mac admin community and easy to read/write with `defaults`. The workflow would be...

1. Write `/var/tmp/.homebrew_pkg_user.plist` with a command like `defaults write /var/tmp/.homebrew_pkg_user HOMEBREW_PKG_USER penny`. A normal thing to do with a Munki [preinstall script](https://github.com/munki/munki/wiki/Pre-And-Postinstall-Scripts) or Jamf policy. Allows an admin to include whatever logic they choose for the user.
2. Install the package.
3. Optionally clean up `/var/tmp/.homebrew_pkg_user.plist`

I think this meets both of our requirements. Also fixed a few edge bugs in postinstall for Intel machines.

File: package/scripts/postinstall [hunk 10]
```diff
@@ -57,7 +57,7 @@ then
   if [[ -f "/usr/local/bin/brew" && -d "/usr/local/Homebrew" ]]
   then
     cp -pRL "${homebrew_directory}/.git" "/usr/local/Homebrew/"
-    mv "${homebrew_directory}/cache_api" "/usr/local/Homebrew/"
+    mv -n "${homebrew_directory}/cache_api" "/usr/local/Homebrew/"
```
Comment (MikeMcQuaid) [inline package/scripts/postinstall:60]: @nstrauss Why is this needed? This directory should never exist already, no?

File: package/scripts/postinstall [hunk 11]
```diff
@@ -79,18 +79,29 @@ fi
...
 fi
 
 # move API cache to ~/Library/Caches/Homebrew
-user_api_cache_dir=~"${logged_in_user}"/Library/Caches/Homebrew/api
+user_home_dir=$(dscl . read /Users/"${homebrew_pkg_user}" NFSHomeDirectory | awk '{ print $2 }')
+user_cache_dir="${user_home_dir}/Library/Caches/Homebrew"
+user_api_cache_dir="${user_cache_dir}/api"
 mkdir -vp "${user_api_cache_dir}"
 mv -v "${homebrew_directory}/cache_api/"* "${user_api_cache_dir}"
+chown -R "${homebrew_pkg_user}:staff" "${user_cache_dir}"
```
Comment (MikeMcQuaid) [inline package/scripts/postinstall:106]:
```suggestion
chown "${homebrew_pkg_user}:staff" "${user_home_dir}/Library/"
chown "${homebrew_pkg_user}:staff" "${user_home_dir}/Library/Caches/"
chown -R "${homebrew_pkg_user}:staff" "${user_cache_dir}"
```

perhaps in case we just created `Library` and/or `Caches` owned by `root`?

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good! Couple more tiny nits and I think we're good to go with this. As mentioned before: that will likely be me squashing and cherry-picking this into another non-fork branch for testing the CI job with all secrets etc.

Thanks again @nstrauss!

File: package/scripts/postinstall (hunk 10 above)
Comment (MikeMcQuaid) [reply package/scripts/postinstall:60]: Ok, might be worth doing `rm -rf` instead to handle that case rather than avoid copying new data?

**Comment by nstrauss** [comment]
> Looks good! Couple more tiny nits and I think we're good to go with this. As mentioned before: that will likely be me squashing and cherry-picking this into another non-fork branch for testing the CI job with all secrets etc.
> 
> Thanks again @nstrauss!

@MikeMcQuaid 🙏Thank you for your help and guidance along the way. I'm happy with where it's at now. Will leave the rest up to you. Otherwise let me know how else I can help.

File: package/scripts/postinstall (hunk 10 above)
Comment (nstrauss) [reply package/scripts/postinstall]: Choosing to remove the `-n` option rather than risk stale data. Tested again and couldn't replicate the same failure path. I think it was originally due to the git safe directory stuff. Seeing as it's an edge case, and maybe not even accessible anymore, fine with going back to bare `mv`. https://github.com/Homebrew/brew/pull/16073/commits/8560827cdb0949a34ff701ea317c89328c4afd88

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock, @nstrauss!
